include/picongpu/particles/collision/binary/RelativisticBinaryCollision.hpp

```diff
@@ -0,0 +1,513 @@
+                                        return 1.0_COLL;
+                                }
+                                else
+                                    return cosXi;
+                            }
+                            else if(s12 < 6.0_COLL)
+                            {
+                                float_COLL a;
+                                if(s12 < 3.0_COLL)
+                                {
+                                    float_COLL s12sq = s12 * s12;
+                                    a = 0.0056958_COLL + 0.9560202_COLL * s12 - 0.508139_COLL * s12sq
+                                        + 0.47913906_COLL * s12sq * s12 - 0.12788975_COLL * s12sq * s12sq
+                                        + 0.02389567_COLL * s12sq * s12sq * s12sq;
+                                    a = 1.0_COLL / a;
+                                }
+                                else
+                                {
+                                    a = 3.0_COLL * math::exp(-1.0_COLL * s12);
+                                }
+                                float_COLL bracket = math::exp(-1.0_COLL * a) + 2.0_COLL * u * std::sinh(a);
+                                float_COLL cosXi = math::log(bracket) / a;
+                                if(cosXi < -1.0_COLL || cosXi > 1.0_COLL)
+                                {
+                                    if(cosXi < -1.0_COLL)
+                                        return -1.0_COLL;
+                                    else
+                                        return 1.0_COLL;
+                                }
+                                else
+                                    return cosXi;
+                            }
+                            else
+                            {
+                                // + 1 in Perez 2012 but cos in [-1,1]. Smilei uses -1.
+                                return 2.0_COLL * u - 1.0_COLL;
+                            }
+                        }
+
+                        /* Calculate the momentum after the collision in the COM frame
+                         *
+                         * @param p momentum in the COM frame
+                         * @param cosXi cosine of the scattering angle
+                         * @param phi azimuthal scattering angle from [0, 2pi]
+                         */
+                        static DINLINE float3_COLL
+                        calcFinalComsMomentum(float3_COLL const p, float_COLL const cosXi, float_COLL const phi)
+                        {
+                            float_COLL sinPhi, cosPhi;
+                            pmacc::math::sincos(phi, sinPhi, cosPhi);
+                            float_COLL sinXi = math::sqrt(1.0_COLL - cosXi * cosXi);
+
+                            // (12) in [Perez 2012]
+                            float3_COLL finalVec;
+                            // TODO: sqrt returns a pmacc complex number. cast to float?
```

Comment:
Why does it return a complex number at all? 
Why is `pmacc::math::abs2(p)` not positive?

Reply:
I don't think that this is really important. It's that the return type for sqrt is comples but the result is never complex. I removed the comment.

Reply:
Could it be that you are using [this `Sqrt`](https://github.com/ComputationalRadiationPhysics/picongpu/blob/a2c46f7e5b5dd9e1103505fc3f4d00ce65f17386/include/pmacc/math/complex/Complex.tpp#L148-L167) not [this `Sqrt`](https://github.com/ComputationalRadiationPhysics/picongpu/blob/454a28efe6a6eb0fa1e0e97031ef2569a1ad313b/include/pmacc/algorithms/math/defines/sqrt.hpp#L39-L42) with [this implementation](https://github.com/ComputationalRadiationPhysics/picongpu/blob/454a28efe6a6eb0fa1e0e97031ef2569a1ad313b/include/pmacc/algorithms/math/doubleMath/sqrt.tpp#L37-L45). @psychocoderHPC and @sbastrakov why is the code here pointing to the complex square root?

Reply:
@PrometheusPi  ccording to my IDE I'm using [this](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/thirdParty/cupla/include/cupla/device/math/Root.hpp). That looks  right.